### ISSUE TITLE
https://github.com/Bersier/physical requires `sbt clean compile` - fails to compile without the clean.

### ISSUE THREAD

**Issue by dwalend** [lampepfl/dotty#15724]
## Compiler version

3.1.3

## Minimized code

It's four files and a build.sbt - not very minimized, but several of us have been able to reproduce it. (If you can wait two weeks when I get a bit of free time I'd be happy to see if I can bracket the behavior. I think reporting this early is best.)

Start with https://github.com/Bersier/physical

## Output (click arrow to expand)

`sbt compile` will finish in maybe 10 seconds, works great, everything compiles. Successive `sbt clean compile` calls work fine.

Make any change - maybe rename something - to Dimensions.scala .(Another person on Scala Users reported problems when adding just a blank line.)

`sbt compile` now takes so long that it seems like an infinite loop. It finishes after perhaps 10 minutes, failing to compile with a lot of excuses despite no interesting changes to the code.

I don't believe the problem is the opaque type in Dimensions because I replaced it with a value class and got the same effect.

**Comment by griggt** [comment]
The project seems to fail compilation with match type reduction errors on the 3.2.0-RCs and latest nightly, so that may need to be addressed first.

**Comment by szymon-rd** [comment]
On `3.2.0-RC2`, it fails with 
```[error] -- [E007] Type Mismatch Error: /.../src/main/scala/types/AdditionalUnits.scala:9:38```
But compiles on 3.1.3. Therefore, marking it as a regression. Please note that there are two issues:
 - A long compilation that is reported by @dwalend 
 - Regression with failing compilation

**Comment by szymon-rd** [comment]
It would be very helpful if you could provide minimization in a free moment :)

**Comment by Bersier** [comment]
The code has grown since this has been reported. [Here](https://github.com/Bersier/physical/tree/7db627630e1837d3bf5ed5ef4c991bfe8e51a243) is a link to the code as it was back then. Frustratingly, I cannot even reproduce the issue at this time because of another bug:

```
sbt thinks that server is already booting because of this exception:
sbt.internal.ServerAlreadyBootingException: java.io.IOException: org.scalasbt.ipcsocket.NativeErrorException: [95] Operation not supported
```

**Comment by Bersier** [comment]
I filed [this ticket](https://github.com/lampepfl/dotty/issues/15926) about the regression (assuming it is the only one).

**Comment by dwalend** [comment]
I was able to (finally) commit some time, and can report some updates:

The part of the problem where the compiler runs for an absurdly long time seems gone in Scala 3.3.0.

The part of the problem where it needs an `sbt clean test` to compile remains. I was able to bracket things a little bit.

With 3 (or 2) dimensions - just Length, Time, and Temperature - things work fine consistently without the clean.

With 4 or more dimensions - Length, Time, Temperature, and Mass - the bug appears.

Procedure is `sbt clean test` , rename `ceil` to `ceilNoClean` in Dimensions.scala , `sbt test` .  With 4  or more dimensions the second compile fails with 

```
[error] -- [E007] Type Mismatch Error: /Users/dwalend/projects/physical/src/main/scala/dimensional/additionalunits/additionalUnits.scala:13:45 
[error] 13 |val electronVolt: Energy = 1.602176634e-19 * joule
[error]    |                                             ^^^^^
[error]    |                            Found:    (y$proxy5 :
[error]    |                              (dimensional.dimension.Dimensions.joule :
[error]    |                                dimensional.dimension.Dimensions.Energy)
[error]    |                             & $proxy10.Energy)
[error]    |                            Required: Double
...
``` as a first error.

If you'd like to use my fork for bracketing and testing, start with my fork's branches:

* https://github.com/dwalend/physical/tree/worksWith3Dimensions - works fine 
* https://github.com/dwalend/physical/tree/brokenWith4Dimensions -  shows the bug

**Comment by dwijnand** [comment]
Fixed with #17180

**Comment by dwalend** [comment]
Thanks!
